Panic in UpdatePathAttrs4ByteAs during specific malformed attribute processing

### Description

I found a reproducible `runtime error: index out of range` panic in GoBGP when it receives a specifically crafted BGP UPDATE message. The panic occurs in the `UpdatePathAttrs4ByteAs` function.
Through testing, I found that the crash seems to be triggered by a combination of two factors:
1.The order of path attributes (AS4_PATH appearing before AS_PATH).
2.A specific length value for the AS4_PATH attribute:01.
### Reproduction Payload

The following hex payload consistently reproduces the panic:
```
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
00 22 02
00 00                  # Withdrawn Routes Length
00 0b                  # Total Path Attribute Length
40 11 01 02            # AS4_PATH (Type 17, Len 1,any val)
40 02 04 ff ff de ad   # AS_PATH (Type 2, Len 4,any val)
```
### Observed Result

```
{"time":"2026-02-01T14:16:04.624079573+08:00","level":"WARN","msg":"the received Update message was treated as withdraw","Topic":"Peer","Key":"192.168.31.195","State":"BGP_FSM_ESTABLISHED","Error":"flags are invalid. attribute type: BGP_ATTR_TYPE_AS4_PATH, expect: TRANSITIVE|OPTIONAL, actual: TRANSITIVE"}
panic: runtime error: index out of range [1] with length 1

goroutine 29 [running]:
github.com/osrg/gobgp/v4/internal/pkg/table.UpdatePathAttrs4ByteAs(0xc00021b680, 0xc000201020)
        /home/runner/work/gobgp/gobgp/internal/pkg/table/message.go:206 +0x14f5
github.com/osrg/gobgp/v4/pkg/server.(*fsmHandler).recvMessageloop(0xc000242e40, {0x10e47f0, 0xc000236c80}, {0x10eea98, 0xc00021b6f0}, 0xc000396540, 0xc0002279d0, 0xc0002005a0?)
        /home/runner/work/gobgp/gobgp/pkg/server/fsm.go:1732 +0xf65
created by github.com/osrg/gobgp/v4/pkg/server.(*fsmHandler).established in goroutine 24
        /home/runner/work/gobgp/gobgp/pkg/server/fsm.go:1805 +0x2ca
```
### Analysis

> According to RFC 6793 Section 6 (Error Handling),A NEW BGP speaker that receives a malformed AS4_PATH attribute in an UPDATE message from an OLD BGP speaker MUST discard the attribute and continue processing the UPDATE message.  The error SHOULD be logged locally for analysis.

I observed a specific behavior when GoBGP handling AS4_PATH attributes set different lengths (type 17):
- When the length is set to 02 or 03, the peer correctly rejects the packet with an Attribute Length Error notification.
- However, when the length is set to 01 (e.g., 40 11 01 02), the packet seems to pass the initial parsing layer and reaches the table update logic, eventually triggering the panic.

It appears the crash happens due to an index shift in msg.PathAttributes at [Line 112]:
```
// Line 112: If AS4_PATH precedes AS_PATH, the deletion causes all subsequent attributes to shift left.
msg.PathAttributes = append(msg.PathAttributes[:as4AttrPos], msg.PathAttributes[as4AttrPos+1:]...)

// Line 206: The stale asAttrPos index is used here.
//The function continues to use the stale index (asAttrPos) to update the AS_PATH. Since the slice length has decreased, accessing the old index leads to a panic.
msg.PathAttributes[asAttrPos] = bgp.NewPathAttributeAsPath(newIntfParams)
```
I suspect that when the AS4_PATH length is 01, the parser might be creating an internal attribute object that is considered "valid enough" to be passed to the update logic, but its existence triggers the deletion-induced index shift.

Regarding the attribute ordering, while RFC 4271 Section 5 suggests that attributes SHOULD be ordered by Type Code, it is not a MUST requirement. Therefore, the implementation would ideally be order-agnostic to handle such cases gracefully.

Following RFC 7606, the attribute sequence or content is considered suboptimal,and the expected behavior is "Attribute Discard" or "Treat-as-withdraw".


### Suggested Fix
- The logic in UpdatePathAttrs4ByteAs needs to account for the index shift caused by the in-place deletion of attributes.
- Ensure the attribute parser does not yield incomplete objects when the attribute length is too short to be valid.

### Config
```
$ cat gobgp-server1.conf
[global.config]
  as = 65001
  router-id = "192.168.31.142"

[[neighbors]]
  [neighbors.config]
    neighbor-address = "192.168.31.195"
    peer-as = 65002
    [neighbors.transport.config]
    passive-mode = true
```
```
$ ./gobgpd --version
gobgpd version 4.2.0
```
```
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.5 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.5 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

--- a/internal/pkg/table/message.go
+++ b/internal/pkg/table/message.go
@@ -110,6 +110,11 @@ func UpdatePathAttrs4ByteAs(logger *slog.Logger, msg *bgp.BGPUpdate) {
 
 	if as4Attr != nil {
 		msg.PathAttributes = append(msg.PathAttributes[:as4AttrPos], msg.PathAttributes[as4AttrPos+1:]...)
+		// Adjust asAttrPos if AS4_PATH was before AS_PATH
+		// (AS4_PATH removal shifts the indices of subsequent attributes)
+		if as4AttrPos < asAttrPos {
+			asAttrPos--
+		}
 	}
 
 	if asAttr == nil || as4Attr == nil {

--- a/internal/pkg/table/message_test.go
+++ b/internal/pkg/table/message_test.go
@@ -706,3 +706,44 @@ func TestMergeV4Withdraw(t *testing.T) {
 		assert.True(t, len(d) < bgp.BGP_MAX_MESSAGE_LENGTH)
 	}
 }
+
+// Test for issue #3308: AS4_PATH appearing before AS_PATH
+// This test ensures that when AS4_PATH (Type 17) appears before AS_PATH (Type 2)
+// in the attributes list, the code correctly handles the index adjustment after
+// AS4_PATH deletion without causing an index out of range panic.
+//
+// before:
+//
+//	as4-path : 400000, 300000, 40001 (Type 17, appears first)
+//	as-path  : 65000, 4000, 23456, 23456, 40001 (Type 2, appears second)
+//
+// expected result:
+//
+//	as-path  : 65000, 4000, 400000, 300000, 40001
+func TestAsPathAs4PathOrdering(t *testing.T) {
+	// Create AS_PATH with AS_TRANS
+	as := []uint16{65000, 4000, bgp.AS_TRANS, bgp.AS_TRANS, 40001}
+	params := []bgp.AsPathParamInterface{bgp.NewAsPathParam(bgp.BGP_ASPATH_ATTR_TYPE_SEQ, as)}
+	aspath := bgp.NewPathAttributeAsPath(params)
+
+	// Create AS4_PATH
+	as4 := []uint32{400000, 300000, 40001}
+	param4s := []*bgp.As4PathParam{bgp.NewAs4PathParam(bgp.BGP_ASPATH_ATTR_TYPE_SEQ, as4)}
+	as4path := bgp.NewPathAttributeAs4Path(param4s)
+
+	// IMPORTANT: Put AS4_PATH before AS_PATH to reproduce issue #3308
+	msg := bgp.NewBGPUpdateMessage(nil, []bgp.PathAttributeInterface{as4path, aspath}, nil).Body.(*bgp.BGPUpdate)
+
+	// This should not panic
+	UpdatePathAttrs4ByteAs(logger, msg)
+
+	// Verify the result
+	assert.Equal(t, len(msg.PathAttributes), 1)
+	assert.Equal(t, len(msg.PathAttributes[0].(*bgp.PathAttributeAsPath).Value), 1)
+	assert.Equal(t, len(msg.PathAttributes[0].(*bgp.PathAttributeAsPath).Value[0].(*bgp.As4PathParam).AS), 5)
+	assert.Equal(t, msg.PathAttributes[0].(*bgp.PathAttributeAsPath).Value[0].(*bgp.As4PathParam).AS[0], uint32(65000))
+	assert.Equal(t, msg.PathAttributes[0].(*bgp.PathAttributeAsPath).Value[0].(*bgp.As4PathParam).AS[1], uint32(4000))
+	assert.Equal(t, msg.PathAttributes[0].(*bgp.PathAttributeAsPath).Value[0].(*bgp.As4PathParam).AS[2], uint32(400000))
+	assert.Equal(t, msg.PathAttributes[0].(*bgp.PathAttributeAsPath).Value[0].(*bgp.As4PathParam).AS[3], uint32(300000))
+	assert.Equal(t, msg.PathAttributes[0].(*bgp.PathAttributeAsPath).Value[0].(*bgp.As4PathParam).AS[4], uint32(40001))
+}

